[Do not merge!] Pseudo PR for first release (#31)
nf-core/lsmquant: docs/usage.md

# nf-core/lsmquant: Usage

## :warning: Please read this documentation on the nf-core website: [https://nf-co.re/lsmquant/usage](https://nf-co.re/lsmquant/usage)

> > _Documentation of pipeline parameters is generated automatically from the pipeline schema and can no longer be found in markdown files._

Author: please keep this part in, to point people to the right place :slightly_smiling_face:

```bash
--input '[path to samplesheet file]'
```

Samplesheet header:

```csv title="samplesheet.csv"
sample_id,img_directory,parameter_file
```

| Column           | Description                                                     |
| ---------------- | --------------------------------------------------------------- |
| `sample_id`      | Custom sample name.                                             |
| `img_directory`  | Full path to the image directory for the sample.                |
| `parameter_file` | Full path to the corresponding parameter_file for the analysis. |

### Single or Multiple samples

The pipeline always takes the samplesheet as input. For processing only one sample, you would only specify one sample in the samplesheet. The samplesheet below shows an example for processing multiple samples with the pipeline.

```csv title="samplesheet.csv"
sample_id,img_directory,parameter_file
TEST1,/path/to/TEST1/,/path/to/params_TEST1.csv
TEST2,/path/to/TEST2/,/path/to/params_TEST2.csv
TEST3,/path/to/TEST3/,/path/to/params_TEST3.csv
```

> If different samples should be processed with the same parameter set specified in the `params.csv`, you can use the same `params.csv` for different samples.

Author: Show also an example samplesheet for that

> ### Parameter file

Author: The whole parameter file will be a bit of a pain to maintain (keep the documentation in sync, validation, etc.), but I don't see another way to do this I am afraid.

Reviewer: i have ideas to reduce it. But i would need to update the MATLAB tools for that. So something for a minor release

> In the `parameter.csv` file you should specify processing parameters for your data and pipeline run. The `CSV` contains specific fields that are needed for the processes to run and only the value column should be modified. You can download a template parameter file [here](../assets/params_template_lsmquant.csv).

Author: ```suggestion
In the `parameter.csv` file you should specify processing parameters for your data and pipeline run for each sample. The `CSV` contains specific fields that are needed for the processes to run and only the value column should be modified. You can download a template parameter file [here](../assets/params_template_lsmquant.csv).
```

An example row is displayed below:

```csv title="params.csv"
Parameter,Value
z_window,5
```

> The individual parameters are explained in the methods [section](###Analysisspecificparameters).

Author: ```suggestion
The individual parameters are explained in the following section.
```

### Analysis specific parameters

> This section describes every parameter that can be set in the `parameter.csv`. In order for the pipeline to run correctly all named parameters need to be present in the parameter file and its recommended to use the provided parameter file (link). Every parameter has a default value that will be set if not otherwise defined in the `parameter.csv`.

Author: ```suggestion
This section describes every parameter that can be set in the `parameter.csv`. In order for the pipeline to run correctly all named parameters need to be present in the parameter file and its recommended to use the [provided parameter file]((../assets/params_template_lsmquant.csv). Every parameter has a default value that will be set if not otherwise defined in the `parameter.csv`.
```

#### Sample specific information

|                        |                                                                                                                                                                          |
| ---------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |

> | `group`                | Group name/id.**Default: TEST;WT;R1**                                                                                                                                    |

Author: ```suggestion
| `group`                | Group name/id. *Default: TEST;WT;R1*                                                                                                                                    |
```

added a space before default and make them italic, to make them easier to parse. same should be done to all other entries here.

| `channel_num`          | Channel id.**Default: C01;C00**                                                                                                                                          |

> | `markers`              | Name of markers present.**Default: topro;ctip2**                                                                                                                         |

Author: ```suggestion
| `markers`              | Name of present markers.**Default: topro;ctip2**                                                                                                                         |
```

> | `position_exp`         | 1x3 string of regular expression specifying image row(y), column(x), slice(z).**Default: [\d*;\d*];Z\d**                                                                 |

Author: ```suggestion
| `position_exp`         | 1x3 string of regular expression specifying image row(y), column(x), slice(z).**Default: [\d\*;\d\*];Z\d**                                                                 |
```

breaks the markdown rendering otherwise: https://nf-co.re/lsmquant/dev/docs/usage/#sample-specific-information

| `resolution`           | Image resolution in um/voxel.**Default: ''**                                                                                                                             |
| `orientation`          | 1x3 string specifying sample orientation. **Default: ail**                                                                                                               |
| `hemisphere`           | "left","right","both","none". **Default: left**                                                                                                                          |

> | `use_processed_images` | false or name of sub-directory in output directory (i.e. aligned, stitched...); Load previously processed images in output directory as input images. **Default: false** |

Author: ```suggestion
| `use_processed_images` | `false` or name of sub-directory in output directory (i.e. aligned, stitched...); Load previously processed images in output directory as input images. **Default: false** |
```

same should be done for all other values, to indicate them more clearly

> | `ignore_markers`       | completely ignore marker from processing steps. **Default: Auto**                                                                                                        |

Author: what does `auto` do? is this not a boolean?

| | |
| ------------- | ---------- |
| `update_z_adjustment` | true, false; Update z adjustment steps with new parameters. Otherwise pipeline will search for previously calculated parameters. **Default: false** |
| `z_positions` | integer or numeric; Sampling positions along adjacent image stacks to determine z displacement. If <1, uses fraction of all images. Set to 0 for no adjustment, only if you're confident tiles are aligned along z dimension. **Default: 0.01** |
| `z_window` | integer; Search window for finding corresponding tiles (i.e. +/-n z positions). **Default: 5** |
| `z_initial` | 1xn_channels-1 integer; Predicted initial z displacement between reference channel and secondary channel. **Default: 0** |

For align by translation
| | |
| ------------- | ---------- |
| `align_stepsize` | integer; Only for alignment by translation. Number of images sampled for determining translations. Images in between are interpolated. **Default: 5** |
| `only_pc` | true, false; Use only phase correlation for registration. This gives only a quick estimate for channel alignment. **Default: false** |

Specific for align by elastix
| | |
| ------------- | ---------- |

> | `align_chunks` | Only for alignment by elastix. Option to align only certain chunks. **Default: ''** |

Author: do make it clearer that these are only applicable to elastix, maybe add a  `elastix_` prefix to all of them?

| `upperThresh`         | 1xn_channels numeric; Upper intensity for rescaling. **Default: ''**                                 |
| `Gamma`               | 1xn_channels numeric; Gamma intensity adjustment. **Default: ''**                                    |

Parameters for background subtraction
| | |
| --------------------------- | -------------- |
| `subtract_background` | true, false. Subtract background (similar to Fiji's rolling ball background subtraction).**Default: false** |
| `nuc_radius` | numeric >= 1; Max radius of cell nuclei along x/y in pixels. Required also for DoG filtering.**Default: 13** |

Difference-of-Gaussian filter
| | |
| --------------------------- | -------------- |
| `DoG_img` | true,false; Apply difference of gaussian enhancement of blobs.**Default: false** |
| `DoG_minmax` | 1x2 numeric; Min/max sigma values to take differences from.**Default: 0.8;2** |
| `DoG_factor` | [0,1]; Factor controlling amount of adjustment to apply. Set to 1 for absolute DoG.**Default: 1** |

Smoothing filters
| | |
| --------------------------- | -------------- |

> | `smooth_img` | 1xn_channels, "gaussian", "median", "guided". Apply a smoothing filter.**Default: false** |

Author: is these really 1xn_channels? and the default value is also not in the range of given values

| `points_file`               | Name of points file to guide registration. **Default: ''**                                                                                              |
| `save_registered_images`    | Whether to save registered images. **Default: true**                                                                                                    |
| `mask_cerebellum_olfactory` | Remove olfactory bulbs and cerebellum from atlas ROI. **Default: true**                                                                                 |

#### Nuclei Detection

|                 |                                                       |
| --------------- | ----------------------------------------------------- |
| `count_method`  | **Default: 3dunet**                                   |
| `int_threshold` | Minimum intensity of positive cells. **Default: 200** |

3-DUnet specific parameters
| | |
| --------------------------- | -------------- |
| `model_file` | Model file name. **Default: ''** |
| `gpu` | Cuda visible device index. **Default: 0** |

> | `chunk_size` | 'Chunk size in voxels. **Default: [112, 112, 32]** |

Author: ```suggestion
| `chunk_size` | Chunk size in voxels. **Default: [112, 112, 32]** |
```

- `wave`
  - A generic configuration profile to enable [Wave](https://seqera.io/wave/) containers. Use together with one of the above (requires Nextflow ` 24.03.0-edge` or later).
- `conda`

>   - This profile is not available for nf-core/lsmquant

Author: would be nice if you would give a quick reason for that

```bash
NXF_OPTS='-Xms1g -Xmx4g'
```

## Methods description

This section provides a detailed explanation for the individual processing steps, based on the original NuMorph toolbox [publication](https://www.sciencedirect.com/science/article/pii/S2211124721012626?via%3Dihub).

### Intensity adjustment

This step performs two types of intensity adjustments to the raw images before tile stitching:

- Shading correction (Option: BaSiC, manual)

- Normalizing intensities between tile stacks

**Shading correction by using BaSiC**

> The Gaussian shape of the light-sheet causes uneven illumination and shading across the y-axis. To correct these effects the tool BaSiC (MATLAB tool for retrospective shading correction) is used. A fraction of all images is used to estimate the flatfiled for each channel. Every image is then divided by the estimated flatfield to normalize illumination. With the parameter `sampling_frequency`, the fraction of images to use for BaSiC, is specified as a decimal. For example setting the `sampling_frequency` to `0.1`, `10%` of all images will be used to estimate the flatfield.

Author: ```suggestion
The Gaussian shape of the light-sheet causes uneven illumination and shading across the y-axis. To correct these effects the tool BaSiC (MATLAB tool for retrospective shading correction) is used. A fraction of all images is used to estimate the flatfield for each channel. Every image is then divided by the estimated flatfield to normalize illumination. With the parameter `sampling_frequency`, the fraction of images to use for BaSiC, is specified as a decimal. For example setting the `sampling_frequency` to `0.1`, `10%` of all images will be used to estimate the flatfield.
```

**Normalizing intensities between tile stacks**

Photo-bleaching and light attenuation can cause differences in brightness between tile stacks. To account for that, the differences in intensities are measured in overlapping regions (vertical and horizontal) of adjacent tiles. Next the adjustment factor $t^{adj}$, based on the 95th percentile of pixel intensities in overlapping regions, is calculated. For this 5% of all images are used.
The final adjustment is applied with the following formula:

> $$I^{adj}(x,y) = (I(x,y) - D)*t^{adj}(x,y) + D $$

Author: ```suggestion
$$
I^{adj}(x,y) = (I(x,y) - D) \cdot t^{adj}(x,y) + D
$$
```

- $I(x,y)$: Original measured image intensities at tile position (x,y)
- $I^{adj}(x,y)$: Adjusted image intensities at tile position (x,y)
- $t^{adj}$: Adjustment factor based on the 95th percentile of intensity differences
- $D$: Darkfield intensity (constant value based on the 5th percentile of pixel intensities across all measured regions)

### Channel alignment

Drifts in sample positions and stage can occur during imaging multiple channels causing spatial misalignment. To correct for these shifts between channels two methods can be chosen:

> - Rigid 2D translation: `align_method`: `translation`

Author: ```suggestion
- Rigid 2D translation: `align_method: translation{:groovy}`
```

> - Nonlinear 3D registration using Elastix : `align_method`: `elastix`

Author: ```suggestion
- Nonlinear 3D registration using Elastix : `align_method: elastix{:groovy}`
```

Both methods expect the nuclear channel as reference, to which all other immunolabeled channels will be aligned to.

> **Rigid 2D translation**

Author: ```suggestion
#### Rigid 2D translation
```

makes it linkable

> This approach estimates first the relative z displacement between the nuclei reference channel and every other channel. Within each tile, a number evenly spaced z slices of the reference channel is chosen by the parameter `z_position`. For every z position, phase corelation is calculated between all images from another channel in a search window (set by `z_window`) and summed up.

Author: ```suggestion
This approach estimates first the relative z displacement between the nuclei reference channel and every other channel. 
Within each tile, a number of evenly spaced z slices of the reference channel is chosen by the parameter `z_position`. 
For every z position, phase corelation is calculated between all images from another channel in a search window (set by `z_window`) and summed up.
```

The z position with the highest image similarity based on intensity correlation defines the inter-channel z displacement

Next, multimodal 2D registration is performed on each slices in the image stack by using MATLAB's Image Processing toolbox, to determine xy translations. Outlier translations are defined as translations that are greater than 3 scaled median absolute deviations within a local window of 10 slices. These outliers are corrected by linear interpolation of adjacent images in the stack.

> **Nonlinear 3D registration using Elastix**

Author: ```suggestion
#### Nonlinear 3D registration using Elastix
```

2. **Normalizing intensities**:
   For comparable brightness and contrast, intensity histogram matching is performed across all channels of the stack.

3. **Generation of foreground mask**:
   A mask for the nuclei channel is generated by using a threshold that limits sampling from the background.

4. **Initial global alignment**:
   An initial 3D translational registration to the full stack is applied.

5. **Local rigid alignment**:
   The stack is subdivided into chunks of 300 slices and a rigid 3D registration on each chunk is performed. This step corrects rotational drift and improves the alignment within local regions.

6. **Nonlinear refinement**:
   A nonlinear B-spline registration is performed on each chunk by using an advanced Mattes mutual information metric. This accounts for xy drift along the z axis. The control point grid of the B-spline transformation are set to be sparse along xy compared to z to balance alignment accuracy and computational cost.

### Iterative image stitching

The iterative 2D stitching procedure to assemble the whole image, consists of two main stages:

- Estimation of z correspondence between tile stacks
- Iterative xy alignment and stitching

> **Estimation of z correspondence between tile stacks**

Author: ```suggestion
#### Estimation of z correspondence between tile stacks
```

> To determine optimal z correspondence for adjacent tiles, a sample of evenly spaced images (set with `z_position`) from within a stack are registered to every z position within a image window (set by `z_window`) of a adjacent stack (vertically and horizontally) by phase correlation. Z correspondences are ranked by the amount of peak correlations among the z positions, where the highest count represent the best correspondence. In addition, the difference between the best and the 2nd best z correlation is taken as a weight, indicating the strength of a correspondence (larger difference = stronger correspondence).

Author: ```suggestion
To determine optimal z correspondence for adjacent tiles, a sample of evenly spaced images (set with `z_position`) from within a stack are registered to every z position within a image window (set by `z_window`) of a adjacent stack (vertically and horizontally) by phase correlation. 
Z correspondences are ranked by the amount of peak correlations among the z positions, where the highest count represent the best correspondence. 

In addition, the difference between the best and the 2nd best z correlation is taken as a weight, indicating the strength of a correspondence (larger difference = stronger correspondence).
```

Finally this results in 4 matrices for a stack representing pairwise horizontal and vertical z displacements and their corresponding weights. To calculate the final z displacement for each tile a minimum spanning tree is used, where displacements are used as vertices and their weights as edges.

> **Iterative xy alignment and stitching**

Author: ```suggestion
#### Iterative xy alignment and stitching
```

Stitching process proceeds with iterative alignment in the x–y plane. The starting point for the iterative stitching along the stack is chosen near the middle of the volume. At this position, all tiles contained sufficient signal above background, defined as at least one standard deviation above the darkfield intensity. The initial translations for each tile are computed using phase correlation, providing a robust estimate of relative positioning. These translations are then refined using the Scale-Invariant Feature Transform (SIFT) algorithm, which improves accuracy by matching distinctive image features across overlapping regions. Stitching begins from the top-left tile to maintain consistent positioning and prevent cumulative shifts along the z-axis. Overlapping areas between tiles are blended using a sigmoidal function, ensuring smooth transitions and preserving image contrast. To handle cases where slices lack sufficient tissue content, shifts greater than five pixels compared to the previous iteration are replaced with the previous iteration’s values.

### Nuclei quantification

> Images are subdivided into patches of 112 × 112 × 32 voxels with an overlap of 16 × 16 × 8 voxels to reduce boundary artifacts. Each patch is then passed to a modified pretrained 3D‑UNet (based on Çiçek et al., 2016 and Isensee et al., 2018) to predict binary nuclei masks. Individual nuclei are obtained via connected‑component analysis, and centroid coordinates are extracted from these components. To prevent duplicate detections introduced by overlapping patches, centroids located closer than half the overlap to a patch border (< 8 pixels in x/y or < 4 pixels in z) are removed under the assumption they will be captured by the neighboring patch. Remaining centroids across all patches are merged using a kd‑tree nearest‑neighbor search, eliminating duplicates within 1.5 voxels of each other to ensure each nucleus is counted exactly once.

Author: feel free to add a link to the papers
